### ISSUE TITLE
BodyPressureSD Dataset readme file

### ISSUE THREAD

**Issue by Tandon-A** [Healthcare-Robotics/BodyPressure#8]
Hello Henry, 

Is there a readme file explaining how the BodyPressureSD dataset is structured and how to use it? 

Best, 
Abhishek

**Comment by Tandon-A** [comment]
Could you also explain what exactly is being done for preparing the depth image? [tensorprep_lib_bp L182-186 ](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/tensorprep_lib_bp.py#L182-L186)

It seems that you are breaking a depth image to be similar to a pressure image in dimensions. Is this depth image then later on built back to a single channel image to be passed to the network? Or does the network take in this representation as input?

**Comment by Tandon-A** [comment]
Could you also explain the labels? [tensorprep_lib_bp L287](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/tensorprep_lib_bp.py#L287)

This is what I have understood: 
gt_markers are the 3D cartesian joint positions, 
body_shape is the beta parameter, 
joint_angles if the theta parameter,
root_xyz_shift is the global translation, 
[g1], [g2] is the gender (which will be passed to SMPL to generate mesh),
body volume, 
body height

Body volume, height will be used for BetaNet loss. 

The model is predicting 89 parameters:
beta (body shape), 
theta (joint angles), 
s (global translation), 
x, y (together define global rotation),
b (distance to the camera), 

But in the labels, there is no part of the global rotation and distance to the camera. 

Am I missing something here?

**Comment by henryclever** [comment]
https://github.com/Healthcare-Robotics/BodyPressure/blob/master/viz_data_only/viz_BodyPressureSD.py

**Comment by henryclever** [comment]
> Hello Henry,
> 
> Is there a readme file explaining how the BodyPressureSD dataset is structured and how to use it?
> 
> Best, Abhishek

No, unfortunately I did not make one. However, there is a script that allows you to visualize what is inside of it. And within that script, it should provide clues as to how to unpack and parse it.

https://github.com/Healthcare-Robotics/BodyPressure/blob/master/viz_data_only/viz_BodyPressureSD.py

Example script: `python viz_BodyPressureSD.py --filenum 2 --pose_num 492 --viz '3D'`

This example script visualizes a specific pose sample within the BodyPressureSD dataset.

**Comment by henryclever** [comment]
> Could you also explain what exactly is being done for preparing the depth image? [tensorprep_lib_bp L182-186 ](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/tensorprep_lib_bp.py#L182-L186)
> 
> It seems that you are breaking a depth image to be similar to a pressure image in dimensions. Is this depth image then later on built back to a single channel image to be passed to the network? Or does the network take in this representation as input?

The purpose of this quad separation of the depth image is to save on space when you are also using pressure images in the input. If you had a single pressure image (64x27) and a single depth image (128x54) and you wanted to stack them together in a single tensor, you'd have to otherwise pad the pressure image with a bunch of zeros or upscale it which is inefficient - requires 8 x 64 x 27 x 100K floats, assuming you are storing the whole dataset in RAM or VRAM. But if you instead break up the depth when iterating over batches in an epoch, you only have to store 5 x 64 x 27 x 100K floats. This makes a difference in Mod2, which already uses lots of VRAM and feeds estimated pressure images back in from Mod1. Prior to making any passes in the network, the depth images are concatenated back together per-batch on L135 here: https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/unpack_depth_batch_lib_bp.py . So this splitting up makes no difference to training or testing the network.

**Comment by henryclever** [comment]
> Could you also explain the labels? [tensorprep_lib_bp L287](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/tensorprep_lib_bp.py#L287)
> 
> This is what I have understood: gt_markers are the 3D cartesian joint positions, body_shape is the beta parameter, joint_angles if the theta parameter, root_xyz_shift is the global translation, [g1], [g2] is the gender (which will be passed to SMPL to generate mesh), body volume, body height
> 
> Body volume, height will be used for BetaNet loss.
> 
> The model is predicting 89 parameters: beta (body shape), theta (joint angles), s (global translation), x, y (together define global rotation), b (distance to the camera),
> 
> But in the labels, there is no part of the global rotation and distance to the camera.
> 
> Am I missing something here?

The global rotation is included in the SMPL joint angles - the first 3 (total 72) -- the rotation of the root SMPL joint, here: `dat['joint_angles'][entry][0:72]`.

~~There is no loss computed over the distance to the camera.~~ In principle, the transform between the bed and the depth camera is fixed, i.e. no free parameters and nothing to learn. As a side note though, this transform is needed when you project the point cloud (or a mesh) into image space -- see L255 here for an example https://github.com/Healthcare-Robotics/BodyPressure/blob/master/viz_data_only/viz_BodyPressureSD.py. But in learning this "should not" be needed since the depth camera is in a fixed frame, constant transform. However, the real world does not behave so ideally. When a heavier person (e.g. 200 lbs) lays on the bed, the mattress sags down substantially, changing the distance between the bed and camera by up to 10cm. EDIT: **The simulated mattress did not sag down as much as the real bed did because there are bed springs underneath the mattress, which is two layers of deformation. My simulation only models the deformation of the mattress, and my data augmentation pipeline at L535 [here](https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/preprocessing_lib_bp.py) models the sag of the bed springs underneath the mattress, where the overall sag is is a function of b.**

~~So the inferred pressure image P^ in Mod2 needs an extra free parameter to adjust for it, or otherwise the loss on PMR does not converge and the inferred pressure is not accurate. There is no ground truth on this distance -- it is learned implicitly from the loss on PMR.~~

**Comment by Tandon-A** [comment]
Thanks, Henry! 

Have one follow-up question. 

In the paper, in appendix C equation 12, the SMPL loss contains terms for x, y, and b. 
Since you mentioned that there is no ground truth on the distance to the camera (b), this is not included in the loss. 
Also, x and y are used to get the global rotation by atan2(y, x) and this is compared with the ground truth (rotation of the root SMPL joint that you mentioned), unlike in equation 12, where predicted y and x are compared with GT y and x. 

Is this correct? Or am I missing something?

**Comment by henryclever** [comment]
> Thanks, Henry!
> 
> Have one follow-up question.
> 
> In the paper, in appendix C equation 12, the SMPL loss contains terms for x, y, and b. Since you mentioned that there is no ground truth on the distance to the camera (b), this is not included in the loss. 

I misspoke. Sorry, it's been a couple years. We did learn the bed height using the loss function, and this is computed from parabolic noise added to the depth image where the location of most sag is at the middle of them mattress, furthest from the steel frame around the perimeter of the mattress. This adjusted distance to the camera (i.e. the bed height ground truth) is a data augmentation parameter generated per image in the function at L535 here - https://github.com/Healthcare-Robotics/BodyPressure/blob/master/lib_py/preprocessing_lib_bp.py. So the amount learned is the distance the bed sags down, measured from 0 (i.e. no sag). I added strikethroughs in my previous post to correct it.

> Also, x and y are used to get the global rotation by atan2(y, x) and this is compared with the ground truth (rotation of the root SMPL joint that you mentioned), unlike in equation 12, where predicted y and x are compared with GT y and x.
> 
> Is this correct? Or am I missing something?

Regarding y and x in the atan2 parameterization: y and x are outputs of the last learned layer of the encoding CNN. The loss on global rotation is computed on these y and x outputs directly following equation 12. Then, atan2 converts them to an angle, which is then input to the SMPL model to determine its global rotation. There is no loss computed on the 3 angles of the SMPL global rotation, because a full global rotation is continuous, but 2pi = 0 in angle space which is not continuous. So it would not be a good idea to learn it from angles. Since there is no loss on the 3 global angles, **Theta** in equation 12 only contains 69 angles (not 72). Alternatively I could have done this with quaternions, which are continuous like atan2 parameterization.

Does this make sense?

We went through the equations many times to ensure they were correct, so I'm pretty sure there are no errors in them in the CVPR version.

**Comment by Tandon-A** [comment]
How do you obtain the GT values for y and x?

**Comment by Tandon-A** [comment]
Understood the part about the loss of bed distance. Thanks for pointing out the line in preprocessing_lib_bp.

**Comment by Tandon-A** [comment]
For the data augmentations, as I understand it, you just added used the preprocessing_add_depth_calnoise function. 
This function is added:
1. white noise
2. pixel dropout
3. region occlusion 
No resizing is applied to the images. And CNN is sent images of size 128 X 54. 
Is this correct? Am I missing anything? 
Did you use any other data augmentations? 

Also, could you let me know if you used any normalization on the images? (mean, std)

**Comment by henryclever** [comment]
I used those 3 augmentations in addition to the bed height augmentation. No resizing was applied except to DMR and PMR which are upscaled by a factor of 2 (64 x 27 --> 128 x 54) before entering the input of a network. The absolute depth images are used, since absolute per pixel value matters and is learned (so no depth normalization). Pressure is normalized however, by body mass, which is explained in the paper.
